packages/fxa-settings/src/pages/Signin/container.tsx

```diff
@@ -552,4 +404,226 @@ const SigninContainer = ({
+          v1Credentials,
+          v2Credentials,
+        };
+      }
+
+      // Determine wrapKb.
+      let wrapKb = '';
+      if (keyFetchToken) {
+        try {
+          const keysResponse = await getWrappedKeys({
+            variables: {
+              input: keyFetchToken,
+            },
+          });
+          wrapKb = keysResponse.data?.wrappedAccountKeys.wrapKB || '';
+        } catch (error) {
+          const uiError = getHandledError(error);
+          if (uiError.error.errno === 104) {
+            // NOOP - If the account is simply 'unverified', then go through normal flow.
+            // The password upgrade can occur later.
+            unverifiedAccount = true;
+          } else {
+            Sentry.captureMessage(
+              'Failure to finish v2 upgrade. Could not get wrapped keys.'
+            );
+            return {
+              ...getHandledError(error),
+              unverifiedAccount,
+              v1Credentials,
+              v2Credentials,
+            };
+          }
+        }
+      }
+
+      // Derive V2 wrapKb and authPW and finalize password reset
+      if (wrapKb && passwordChangeToken) {
+        try {
+          v2Credentials = await getCredentialsV2({
+            password,
+            clientSalt:
+              credentialStatusData.data?.credentialStatus.clientSalt ||
+              createSaltV2(),
+          });
+
+          const kB = unwrapKB(wrapKb, v1Credentials.unwrapBKey);
+          const keys = await getKeysV2({
+            kB,
+            v1: v1Credentials,
+            v2: v2Credentials,
+          });
+
+          // Will always return an empty payload on success, because no session token is provided.
+          await passwordChangeFinish({
+            variables: {
+              input: {
+                passwordChangeToken: passwordChangeToken,
+                authPW: v1Credentials.authPW,
+                wrapKb: keys.wrapKb,
+                authPWVersion2: v2Credentials.authPW,
+                wrapKbVersion2: keys.wrapKbVersion2,
+                clientSalt: v2Credentials.clientSalt,
+              },
+            },
+          });
+        } catch (error) {
+          Sentry.captureMessage(
+            'Failure to finish v2 upgrade. Could not finish password change.'
+          );
+          return {
+            ...getHandledError(error),
+            unverifiedAccount,
+            v1Credentials,
+            v2Credentials,
+          };
+        }
+      }
+    } else if (credentialStatusData.data?.credentialStatus.clientSalt) {
+      v2Credentials = await getCredentialsV2({
+        password,
+        clientSalt: credentialStatusData.data?.credentialStatus.clientSalt,
+      });
+    }
+  }
+
+  return {
+    unverifiedAccount,
+    v1Credentials,
+    v2Credentials,
+    error: undefined,
+  };
+}
+
+/**
+ * Attempts to sign in with v2Credentials if available else fallback to v1credentials for sign in.
+ **/
+export async function trySignIn(
```

Comment:
Should this be moved to `signin/utils` or its own separate file?

Reply:
I was also wondering about this... I can move it over to utils.

Reply:
Oh, this makes some of the mocking weird. I will leave it like this for now, and address this in [FXA-9503](https://mozilla-hub.atlassian.net/browse/FXA-9503).